File: 1. Fundamentals/validParentheses.js
const isValid = function(s) {
    let open = {},
        close = {};
    [
        ['(', ')'],
        ['{', '}'],
        ['[', ']'],
    ].forEach(([o, c]) => {
        open[o] = c;
        close[c] = o;
    });

    console.log(open, close);

    let stack = [];
    for (let i = 0; i < s.length; i++) {
        if (open.hasOwnProperty(s[i])) {
            stack.push(s[i]);
        } else if (close.hasOwnProperty(s[i])) {
            if (open[stack.pop()] !== s[i]) return false;
        }
    }

    return stack.length === 0;
};

File: 1. Fundamentals/valid_parentheses.py
def isValid(self, s: str) -> bool:
  open = {}
  close = {}
  for a in [['(', ')'], ['{', '}'], ['[', ']']]:
    open[a[0]] = a[1]
    close[a[1]] = a[0]
  print(open, close)
  stack = []
  for c in s:
    if (c in open):
      stack.append(c)
    elif (c in close):
      if(len(stack) == 0 or stack.pop() != close[c]):
        return False
    else:
      return False

  return len(stack) == 0
